README
EvoCompX: evolution and competition across space

Version 0.2, Sept 2009
Emma Goldberg <eeg@umd.edu>

This program implements the model of Case and Taper (2000, AmNat), with some
extensions.  In particular, any number of species can be used, and competition
coefficients can be specified.  The phenotype-independent hybridization from
Goldberg and Lande (2006) is also included.  See the end of this file for major
limitations.


Step 1: Compile the program
---------------------------

On Linux/Unix-like systems, probably Mac OS X, and maybe Windows with cygwin,
compilation from a terminal ($ is the prompt) is as simple as:
   $ cd src/
   $ make
If you use an IDE or some other environment, you're on your own.

As a first check, run the executable without any options:
   $ ./EvoCompX
You should see a message saying that input parameters are needed.


Step 2: Run the example
---------------------------

Move to the directory with input examples:
   $ cd ../input/

Take a look at the file params.in.  It includes all the parameters that can be
specified.  The notation mostly follows Case and Taper (2000).  Some of the
parameters have default values, but for most of them, you'll get a warning
message if any are unspecified.

num_species:      need at least 2 species, and currently no more than 20
r, K:             parameters of the logistic growth model
h2:               heritability
V_s, V_p, V_u:    variances of the stabilizing selection, phenotypic, and 
                    competition functions
alpha_file:       a file containing the matrix of competition coefficients,
                    alpha[row i][col j] = alpha_ij = effect of j on i;
                    values should be separated by whitespace
beta:             frequency-independent hybridization
delta:            probability of dispersal to a neighboring cell
space_size:       number of spatial cells, currently no more than 1000
opt_slope:        the optimum phenotype function is currently a line with 0 
                    intercept and this slope [theta(x)]
start_t:          the starting time (useful to change if you want to continue a
                    previous run)
stop_t:           the stopping time; the number of time steps to be run is 
                    stop_t - start_t
record_interval:  after this many time steps, the results are written to files
initial_num, initial_zbar:  files with initial conditions

Initial conditions can be completely arbitrary.  The file specified by
initial_num should contain the initial abundances, with one column per species
and one row per spatial cell.  Similarly for initial_zbar and the initial mean
phenotypes.  These can be written by hand, or generated with the make_initial.py
script included or any other script you want to write (e.g. randomly tossing
individuals onto the landscape).

To run the included three-species example:

Generate the initial condition files num.in and zbar.in:
   $ ./make_initial.py

Generate a file with a competition coefficient matrix, alpha.in:
   $ ./make_alpha.py 0.5
(If a number is specificed [0.5 here], that will be the value of all the a_ij
elements.  If no number is given, random values will be chosen.)

Run the program, generating several output files:
   $ ../src/EvoCompX params.in

You now have:
   time.dat: the times at which the system state was recorded
   num1.dat, num2.dat, num3.dat: the abundances of each species at each location
      and each time of record; one line per time, one column per spatial cell
   zbar1.dat, zbar2.dat, zbar3.dat: same for the mean phenotypes of each species
   num_final.dat, zbar_final.dat: the state of the system at the last time; the
      format of each file is the same as for the input files, so it is easy to
      continue runs

As an example of how to visualize at the output, see output/plot_results.R:
   $ mv *.dat ../output/
   $ cd ../output/
   $ R --no-save < plot_results.R
The file generated is results.pdf.


Step 3: General use
---------------------------

All options can be controlled through the input parameter file, an example of
which is input/params.in.  Such a file must always be the first argument.
Additionally, any of the parameters can be specified on the command line (with
no spaces on either side of the = sign), and those values will override what's 
in the input file.  This can be useful when using a script to run many 
scenarios.  For example:
   $ ../src/EvoCompX params.in opt_slope=0.4 delta=0.01

Hugely increasing the number of species, number of spatial cells, or elapsed
time could lead to slow run times.  So best to increase those just gradually.
If more than 20 species or 1000 cells are needed, the space array must be
allocated dynamically.

Here are some of the current limitations in the code.  All of these can be
changed if that would be useful.

(a) The value for a given parameter is the same for each species, e.g. all
species have the same K.

(b) The value for a given parameter is the same everywhere in space.

(c) The maximum dispersal distance is one cell per generation.  I experimented
with arbitrary dispersal kernels, but that got messy.  A simpler alternative is
to let nearest-neighbor dispersal happen a specified number of times per
generation.

(d) Dispersal is symmetric: the probability of moving left is the same as the
probability of moving right.

(e) Competition is symmetric by phenotype.  To change that (e.g. so larger-z
individuals have a larger effect on smaller-z individuals than vice versa,
regardless of their species), one could use a different competition 
distribution (like Case and Taper's kappa).

(f) The optimum phenotype function, theta(x), so far is just a straight line,
but it could be made into anything.  Modifications can be hard-coded in
optimum.c, or a few general cases can be specified there and appropriate
parameters created.

(g) Selection must be weak, meaning that the cline can't be too steep,
dispersal too far, or the selection function too narrow.  Changing this
requires obtaining the Wrightian fitness numerically rather than through the
current analytical approximation.  This would slow the analysis down a bit, but
it is probably worth doing.


Disclaimer
---------------------------

My older, two-species version of this code was well-tested and gave every
indication of working properly.  I have only tested this multi-species version
briefly -- it appears to work fine, but please keep an eye out for weird
behavior (even with two species, since some of the code had to be re-written).


CHANGES
0.2 (27 Sept 2009)
------------------

Added the matrix of competition coefficients, to allow different strengths of
inter- and intraspecific competition.

0.1 (24 Sept 2009)
------------------

Case and Taper (2000) model + many species allowed + phenotype-independent
hybridization.  Keyvalue for user input.


src/input.c
/**********************************************************************
 * Copyright 2009 Emma Goldberg
 * 
 * This file is part of EvoCompX.
 * 
 * EvoCompX is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * EvoCompX is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with EvoCompX.  If not, see <http://www.gnu.org/licenses/>.
 *********************************************************************/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "input.h"
#include "keyvalue.h"
#include "landscape.h"


/***
 * user input of parameters
 ***/


/*** create a Params structure to hold the input parameter values ***/
Params *NewParams()
{
	Params *p;
	p = (Params *) malloc(sizeof(Params));

	if (p == NULL)
	{
		fprintf(stderr, "NewParams malloc failed\n");
		exit(1);
	}

	return p;
}

void FreeParams(Params *params)
{
	free(params);
}


/*** read in the user's input parameter values ***/
int AcquireParams(struct KeyValue *kv, Params *parameters)
{
	parameters->num_sp = getKeyValueint(kv, "num_species");
     if (parameters->num_sp == KV_INTERR)
	{
		fprintf(stderr, "need to specify number of species, num_species\n");
		return -1;
	}
     if (parameters->num_sp < 2 || parameters->num_sp > MAX_NUM_SP)
	{
		fprintf(stderr, "need 2 <= num_species <= %d\n", MAX_NUM_SP);
		return -1;
	}

	/*** biology ***/

	parameters->r = getKeyValuedouble(kv, "r");
     if (parameters->r == KV_FLOATERR)
	{
		fprintf(stderr, "need to specify growth rate, r\n");
		return -1;
	}

	parameters->K = getKeyValuedouble(kv, "K");
     if (parameters->K == KV_FLOATERR)
	{
		fprintf(stderr, "need to specify carrying capacity, K\n");
		return -1;
	}

	parameters->h2 = getKeyValuedouble(kv, "h2");
     if (parameters->h2 == KV_FLOATERR)
	{
		fprintf(stderr, "need to specify heritability, h2\n");
		return -1;
	}

	parameters->V_s = getKeyValuedouble(kv, "V_s");
     if (parameters->V_s == KV_FLOATERR)
	{
		fprintf(stderr, "need to specify variance of stabilizing selection, "
		        "V_s\n");
		return -1;
	}

	parameters->V_p = getKeyValuedouble(kv, "V_p");
     if (parameters->V_p == KV_FLOATERR)
	{
		fprintf(stderr, "need to specify phenotypic variance, V_p\n");
		return -1;
	}

	parameters->V_u = getKeyValuedouble(kv, "V_u");
     if (parameters->V_u == KV_FLOATERR)
	{
		fprintf(stderr, "need to specify variance of competition function, "
		        "V_u\n");
		return -1;
	}

	parameters->alpha_file = getKeyValuestring(kv, "alpha_file");
     if (parameters->alpha_file == 0)
	{
		fprintf(stderr, "competition matrix not specified, using 1\n");
	}

	parameters->beta = getKeyValuedouble(kv, "beta");
     if (parameters->beta == KV_FLOATERR)
	{
		/* default is no hybridization */
		parameters->beta = 0;
	}

	parameters->delta = getKeyValuedouble(kv, "delta");
     if (parameters->delta == KV_FLOATERR)
	{
		fprintf(stderr, "need to specify dispersal probability, delta\n");
		return -1;
	}

	/*** landscape ***/

	parameters->space_size = getKeyValueint(kv, "space_size");
     if (parameters->space_size == KV_INTERR || parameters->space_size < 1)
	{
		fprintf(stderr, "valid space_size not specified, using 100\n");
		parameters->space_size = 100;
	}

	parameters->opt_slope = getKeyValuedouble(kv, "opt_slope");
     if (parameters->opt_slope == KV_FLOATERR)
	{
		fprintf(stderr, "need to specify slope of optimum phenotype, "
			    "opt_slope\n");
		return -1;
	}

	parameters->initial_num = getKeyValuestring(kv, "initial_num");
     if (parameters->initial_num == 0)
	{
		fprintf(stderr, "need to specify initial abundances, initial_num\n");
		return -1;
	}

	parameters->initial_zbar = getKeyValuestring(kv, "initial_zbar");
     if (parameters->initial_zbar == 0)
	{
		fprintf(stderr, "need to specify initial phenotypes, initial_zbar\n");
		return -1;
	}

	/*** record-keeping ***/

	parameters->start_t = getKeyValueint(kv, "start_t");
     if (parameters->start_t == KV_INTERR)
	{
		fprintf(stderr, "valid start_t not specified, using 0\n");
		parameters->start_t = 0;
	}

	parameters->stop_t = getKeyValueint(kv, "stop_t");
     if (parameters->stop_t == KV_INTERR || 
	    parameters->stop_t < parameters->start_t)
	{
		fprintf(stderr, "valid stop_t not specified, using start_t + 1000\n");
		parameters->stop_t = parameters->start_t + 1000;
	}

	parameters->record_interval = getKeyValueint(kv, "record_interval");
     if (parameters->record_interval == KV_INTERR || 
	    parameters->record_interval <= 0)
	{
		fprintf(stderr, "valid record_interval not specified, using elapsed "
		                "time/10\n");
		parameters->record_interval = 
			(parameters->stop_t - parameters->start_t)/10;
	}

	return 0;		/* returns -1 elsewhere if there's an error */
}


Params *GetParams(int argc, char *argv[])
{
	struct KeyValue *kv;
	char k[100], v[100];
	Params *parameters;
	int i, j;

	/* open the user's file */

	if (argc < 2)
	{
		fprintf(stderr, "\n           *** This is EvoCompX ver 0.2 ***\n"
		     "  Input parameters are required.  Please see the README and\n"
		     "  example params.in file for usage examples and options.\n\n");
		exit(1);
	}

	parameters = NewParams();
	kv = loadKeyValue(argv[1]);
	if (kv == 0)
	{
		fprintf(stderr, "unable to load parameters file\n");
		exit(1);
	}

	/* overwrite parameter values with those specified on the command line */

     if (argc > 2) for(i = 2; i < argc; i++)
	{
		if (argv[i][0] == '=')
		{
			fprintf(stderr, "Warning -- option begins with = "
					"(be sure to use option=value, without spaces)\n");
		}
		for (j = 0; argv[i][j] != 0; j++)
			if (argv[i][j] == '=')
				argv[i][j] = ' ';
		if (sscanf(argv[i], "%s %s", k, v) != 2)
			continue;

		j = KeyValuekeyindex(kv, k);
		if (j < 0)
			KeyValueaddparm(kv, k, v);
		else
		{
			free(kv->value[j]);
			kv->value[j] = strdup(v);
		}
	}

	if (AcquireParams(kv, parameters) == -1)
	{
		fprintf(stderr, "unable to proceed -- "
				"not all required parameters specified\n");
		exit(1);
	}

	return parameters;
}
